### PR TITLE
Improve error messages.

### PR DESCRIPTION
**Pull request by stuartarchibald**
* Compact format
* Traceback interleaving
* Exception "eating" via disappearance/clobbering is fixed.

### CODE REVIEW

File: numba/core/types/functions.py [hunk 1]
```diff
@@ -202,21 +356,63 @@ def get_impl_key(self, sig):
...
+        literal_e = None
+        nonliteral_e = None
+        _DEBUG = False
+
+
         # Try with Literal
         try:
             out = template.apply(args, kws)
-        except Exception:
+        except Exception as exc:
```
Comment (sklam) [inline numba/core/types/functions.py:367]: need to let `errors.ForceLiteralArg` propagate up

File: numba/core/types/functions.py [hunk 2]
```diff
@@ -202,21 +356,63 @@ def get_impl_key(self, sig):
...
+        unliteral_kws = {k: unliteral(v) for k, v in kws.items()}
+        skip = unliteral_args == args and kws == unliteral_kws
+
+        # If the above template application failed and the non-literal args are
+        # different to the literal ones, try again with literals rewritten as
+        # non-literals
+        if not skip and out is None:
+            try:
+                out = template.apply(unliteral_args, unliteral_kws)
+            except Exception as exc:
```
Comment (sklam) [inline numba/core/types/functions.py:383]: need to let `errors.ForceLiteralArg` propagate up

**Comment by stuartarchibald** [comment]
I think there's something broken in testing again, this PR https://github.com/numba/numba/pull/5701 "passed", this one https://github.com/numba/numba/pull/5725, which is the same, but rebased on master, failed. However, the things it's failing on are not new, which implies something weird has happened.

**Comment by sklam** [comment]
RE: https://github.com/numba/numba/pull/5722#issuecomment-630722088

That means testing is still not deterministic. There are still holes in the test slicing.

**Comment by sklam** [comment]
With that problem in testing, i will give it a cpu smoke test in the farm.

**Comment by stuartarchibald** [comment]
> RE: [#5722 (comment)](https://github.com/numba/numba/pull/5722#issuecomment-630722088)
> 
> That means testing is still not deterministic. There are still holes in the test slicing.

May be so, probably need another go at trying to sort that out.

**Comment by stuartarchibald** [comment]
> With that problem in testing, i will give it a cpu smoke test in the farm.

Yeah, I think it'd be wise to do that anyway.

**Comment by sklam** [comment]
`numba_smoketest_cpu_37`

**Comment by sklam** [comment]
Seems like smoketest are affected by entry-point issues. Otherwise, it looks like it has pased.

File: numba/core/types/functions.py [hunk 3]
```diff
@@ -1,58 +1,202 @@
...
+                errstr = _err_reasons['no_explicit_sig']
+                if hasattr(template, '_overload_func'):
+                    if hasattr(template._overload_func.outer, 'signatures'):
+                        sigs = template._overload_func.outer.signatures
+                        if sigs:
+                            sstr = [' ' * (ldepth + 1) + "* " + fn_name + str(x)
+                                    for x in sigs]
+                            lsigs = '\n'.join(sstr)
+                            rtempl = _reason_template['no_match_explicit_sig']
+                            errstr = rtempl.format(lsigs)
```
Comment (stuartarchibald) [inline numba/core/types/functions.py:174]: This is for declarative typing and can probably be removed.

File: numba/core/types/functions.py [hunk 4]
```diff
@@ -1,58 +1,202 @@
...
+            source_file, source_line = self._get_source_info(source_fn)
+            largstr = argstr if err.literal else nolitargstr
+            msgbuf.append(_termcolor.errmsg(
+                _wrapper(_overload_template.format(nduplicates=nduplicates,
+                                                  function=source_fn.__name__,
+                                                  file=source_file,
+                                                  line=source_line,
+                                                  args=largstr),
+                ldepth + 1)))
+            if error is None:
```
Comment (sklam) [inline numba/core/types/functions.py:164]: i guess this whole branch can be removed in this PR.

File: numba/core/types/functions.py [hunk 5]
```diff
@@ -202,21 +356,67 @@ def get_impl_key(self, sig):
...
+                        bt = traceback.format_exception(type(error), error,
+                                                        error.__traceback__)
+                    else:
+                        bt = [""]
+                    nd2indent = '\n{}'.format(2 * indent)
+                    errstr += _termcolor.reset(nd2indent +
+                                               nd2indent.join(bt_as_lines))
+                    return _termcolor.reset(errstr)
+            else:
+                add_bt = lambda X: ''
```
Comment (sklam) [inline numba/core/types/functions.py:407]: this branch seems unreachable.

File: numba/core/extending.py [hunk 6]
```diff
@@ -54,7 +54,8 @@ def generic(self):
 _overload_default_jit_options = {'no_cpython_wrapper': True}
 
 
-def overload(func, jit_options={}, strict=True, inline='never'):
+def overload(func, jit_options={}, strict=True, inline='never',
+             signatures=None):
```
Comment (sklam) [inline numba/core/extending.py:58]: The signatures arguments can be removed for this PR.

File: numba/core/types/functions.py (hunk 3 above)
Comment (sklam) [reply numba/core/types/functions.py:174]: let's remove this for now.

File: numba/core/types/functions.py (hunk 4 above)
Comment (sklam) [reply numba/core/types/functions.py:164]: this is the same as https://github.com/numba/numba/pull/5722#discussion_r430755865

**Comment by sklam** [comment]
retry `numba_smoketest_cpu_41`

**Comment by sklam** [comment]
buildfarm failed but probably unrelated; i.e entry_points test failure on windows, and `numba -s` failures

File: numba/core/types/functions.py (hunk 5 above)
Comment (stuartarchibald) [reply numba/core/types/functions.py:407]: if `DEVELOPER_MODE` is Falseish it always executes, it's the default and provides "no backtrace"?

File: numba/core/extending.py (hunk 6 above)
Comment (stuartarchibald) [reply numba/core/extending.py]: done

File: numba/core/types/functions.py (hunk 4 above)
Comment (stuartarchibald) [reply numba/core/types/functions.py]: done

File: numba/core/types/functions.py (hunk 3 above)
Comment (stuartarchibald) [reply numba/core/types/functions.py]: done

File: numba/core/types/functions.py [hunk 7]
```diff
@@ -202,21 +356,67 @@ def get_impl_key(self, sig):
 
     def get_call_type(self, context, args, kws):
         template = self.template(context)
-        e = None
+        literal_e = None
+        nonliteral_e = None
+        _DEBUG = False
```
[the remark this answers is not in the document]
Comment (stuartarchibald) [reply numba/core/types/functions.py]: done

File: numba/core/types/functions.py (hunk 1 above)
Comment (stuartarchibald) [reply numba/core/types/functions.py:348]: done

File: numba/core/types/functions.py (hunk 2 above)
Comment (stuartarchibald) [reply numba/core/types/functions.py:366]: done

**Pull request merged**